### PR TITLE
docs: openapi v3 decorators

### PR DESCRIPTION
**Pull request by jannyHou**
connect to https://github.com/strongloop/loopback-next/issues/753

## Checklist

- [x] `npm test` passes on your machine(running)
- ~[ ] New tests added or existing tests modified to cover all changes~
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] Related API Documentation was updated
- ~[ ] Affected artifact templates in `packages/cli` were updated~
- ~[ ] Affected example projects in `packages/example-*` were updated~

### CODE REVIEW

File: docs/site/Decorators.md [hunk 1]
```diff
@@ -132,56 +132,181 @@ accordingly or do a composition of them:
   For more usage, refer to [Routing to Controllers](controllers.htm#routing-to-controllers)
 
 ### Parameter Decorator
+ 
+  Syntax: see [API document](https://github.com/strongloop/loopback-next/blob/master/packages/openapi-v3/src/parameter-decorator.ts#L18-L29)
```
Comment (bajtos) [inline docs/site/Decorators.md]: Did you perhaps mean `API documentation`?

Also, could you please point to http://apidocs.loopback.io instead of TypeScript source?

If we decide to keep the link to a ts file, then please make sure the link points to a specific revision, not the `master` branch. Otherwise the link will point to wrong line numbers soon, as the code will evolve.

File: docs/site/Decorators.md [hunk 2]
```diff
@@ -132,56 +132,181 @@ accordingly or do a composition of them:
...
   ```
 
-  In the first example, we apply multiple decorators to a single declaration.
-  The order in which the decorators are called is important because a `@param`
-  decorator must be applied after an operation decorator.
-  To learn more about TypeScript decorator composition,
-  refer to [TypeScript Decorator Documentation](https://www.typescriptlang.org/docs/handbook/decorators.html)
+  You can find the specific usage in [Writing Controller methods](controller.htm#writing-controller-methods)
+
+  *We are supporting parameter location 'cookie' in the feature, track the feature in story*
```
Comment (bajtos) [inline docs/site/Decorators.md]: _The parameter location `cookie` is not supported yet, see https://github.com/strongloop/loopback-next/issues/997._

File: docs/site/Decorators.md [hunk 3]
```diff
@@ -132,56 +132,181 @@ accordingly or do a composition of them:
...
+    description: 'a user',
+    required: true,
+    content: {
+      'application/json': {...schemaSpec},
+      'application/text': {...schemaSpec},
+    },
+  }
+  ```
+
+  When the type of decorated argument can be described as a schema, you can define
```
Comment (shimks) [inline docs/site/Decorators.md]: This sentence needs to be reworded to something like this: `When the decorated argument is a custom type or a model, make sure to decorate the model class with '@model' and '@property'`

File: docs/site/Decorators.md [hunk 4]
```diff
@@ -132,56 +132,181 @@ accordingly or do a composition of them:
...
+  class User {
+    @property()
+    firstname: string
+    @property()
+    lastname: string
+    @property()
+    address: Address
+  }
+  ```
+  *About decorating models and the corresponding OpenAPI schema, please check*
```
Comment (shimks) [inline docs/site/Decorators.md]: `to learn more about decorating ...`

File: docs/site/Decorators.md [hunk 5]
```diff
@@ -132,56 +132,181 @@ accordingly or do a composition of them:
...
+  *[model decorators](#model-decorators)*
+
+  Then a typical usage of `@requestBody` is specifying a parameter's type as the model class,
```
Comment (shimks) [inline docs/site/Decorators.md]: `A typical usage of '@requestBody' is specifying a parameter as a requestBody by decorating it` or something along that line.

File: docs/site/Decorators.md [hunk 6]
```diff
@@ -132,56 +132,181 @@ accordingly or do a composition of them:
...
+  class MyController {
+    @put('/Users/{id}')
+    async replaceUser(
+      @param.path.string('id') id: string,
+      @requestBody() user: User
+    ) {}
+  }
+  ```
+
+  For the simplest use case, you can leave the input of `@requestBody` empty, we automatically
```
Comment (shimks) [inline docs/site/Decorators.md]: `... empty since we automatically ...`

File: docs/site/Decorators.md [hunk 7]
```diff
@@ -132,56 +132,181 @@ accordingly or do a composition of them:
...
+  detect the type of `user`, generate the corresponding schema for it.
```
Comment (shimks) [inline docs/site/Decorators.md]: `... type of 'users' and generate the ...`

File: docs/site/Decorators.md [hunk 8]
```diff
@@ -132,56 +132,181 @@ accordingly or do a composition of them:
...
+    async replaceUser(
+      @param.path.string('id') id: string,
+      @requestBody() user: User
+    ) {}
+  }
+  ```
+
+  For the simplest use case, you can leave the input of `@requestBody` empty, we automatically
+  detect the type of `user`, generate the corresponding schema for it. 
+  And the default content type is set to be `application/json`.
```
Comment (shimks) [inline docs/site/Decorators.md]: `The default content type ...` (no `And`)

File: docs/site/Decorators.md [hunk 9]
```diff
@@ -132,56 +132,181 @@ accordingly or do a composition of them:
...
+          'application/json': UserSchema
+        } 
+      }) user: User
+    ) {}
+  }
+  ```
 
-  So an example would be `@param.query.number('offset')`.
-  You can find the specific usage in [Writing Controller methods](controller.md#writing-controller-methods)
+  *We are supporting more `@requestBody` shorts in the future, track the feature in story*
```
Comment (shimks) [inline docs/site/Decorators.md]: `shorts`?

File: docs/site/Decorators.md (hunk 1 above)
Comment (jannyHou) [reply docs/site/Decorators.md]: @bajtos oops a typo, there is a bug of `ts` doc parsing function:

Given a function and namespace with the same name, it only parses the namespace, but could not detect the function. That's why I point it to the code as a workaround.

File: docs/site/Decorators.md [hunk 10]
```diff
@@ -132,56 +132,181 @@ accordingly or do a composition of them:
...
+      @param(pageSizeSpec) pageSize?: number,
+    ) {}
+  }
   ```
 
-  or
+  Writing the whole parameter specification is tedious, we create shortcuts to define params with pattern `@param.${in}.${type}(${name})`:
+
+  - in: the parameter location, one of the following values: `query`, `header`, `path`.
+  - type: a [common name of OpenAPI primitive data type](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#data-types),
```
[the remark this answers is not in the document]
Comment (jannyHou) [reply docs/site/Decorators.md]: @shimks could you elaborate more about which part confuses you?

[the remark this answers is not in the document]
Comment (jannyHou) [reply docs/site/Decorators.md]: I moved `take specifying a data get from query....` to next paragraph, see if it improves :)

File: docs/site/Decorators.md [hunk 11]
```diff
@@ -132,56 +132,181 @@ accordingly or do a composition of them:
...
+  - type: a [common name of OpenAPI primitive data type](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#data-types), 
+    take specifying a data get from query as an example, a list of available shortcuts can
+    be found in [API document](http://apidocs.loopback.io/@loopback%2fopenapi-v3/#param.query).
+  - name: name of the parameter, should be a `string`, 
 
+  An equivalent example using shortcut decorator would be:
+  
   ```ts
-    // example 2: decorator `@param` applied on parameter level
     class MyController {
```
[the remark this answers is not in the document]
Comment (jannyHou) [reply docs/site/Decorators.md]: `UserController`?

File: docs/site/Decorators.md [hunk 12]
```diff
@@ -132,56 +132,181 @@ accordingly or do a composition of them:
...
-      @get('/')
+      @get('/Pets/{category}')
```
[the remark this answers is not in the document]
Comment (jannyHou) [reply docs/site/Decorators.md]: @shimks is it a known naming convention? Maybe I am wrong, but in explorer, a model's endpoints always start with a upper case(first letter) model name, see https://loopback.io/doc/en/lb3/Use-API-Explorer.html#run-api-explorer

File: docs/site/Decorators.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (dhmlau) [reply docs/site/Decorators.md]: should it be `the common name` instead of `a common name`?

File: docs/site/Decorators.md (hunk 4 above)
Comment (dhmlau) [reply docs/site/Decorators.md]: suggest to remove `please`.

File: docs/site/Decorators.md (hunk 3 above)
Comment (jannyHou) [reply docs/site/Decorators.md]: @shimks Hmm...I would rephrase "make sure" to "we suggest", providing a schema spec without a model class is still an option...

File: docs/site/Decorators.md (hunk 5 above)
Comment (jannyHou) [reply docs/site/Decorators.md]: @shimks  I still prefer to keep "then" :)

There is some code snippet in between, but the context is:
> When the decorated argument is a custom type or a model, we recommend you define 
  a model class with `@model` and `@property`. Then a typical usage of `@requestBody` is specifying a parameter's type as the model class, and applying the decorator on it.

File: docs/site/Decorators.md [hunk 13]
```diff
@@ -132,56 +132,181 @@ accordingly or do a composition of them:
...
+    @property()
+    firstname: string
+    @property()
+    lastname: string
+    @property()
+    address: Address
+  }
+  ```
+  *About decorating models and the corresponding OpenAPI schema, please check*
+  *[model decorators](#model-decorators)*
```
Comment (dhmlau) [inline docs/site/Decorators.md]: missing `.` for the end of the sentence.

File: docs/site/Decorators.md [hunk 14]
```diff
@@ -132,56 +132,181 @@ accordingly or do a composition of them:
...
+  Syntax: see [API document](https://github.com/strongloop/loopback-next/blob/master/packages/openapi-v3/src/parameter-decorator.ts#L18-L29)
 
-  Syntax: [`@param(paramSpec: ParameterObject)`]()
+  Shortcut Pattern: `@param.${in}.${type}(${name})`
 
-  Optional Pattern: `@param.${in}.${type}(${name})`
-
-  `@param` can be applied to method itself or specific parameters, it describes
-  an input parameter of a Controller method.
+  `@param` is applied to method parameters to generate OpenAPI parameter specification for it.
```
[the remark this answers is not in the document]
Comment (jannyHou) [reply docs/site/Decorators.md]: section "parameter decorator" and "requestBody decorator" are under "Route decorators", which says they are for controllers. But true, the sections is far from their parent, I added keyword "controller" for both.

File: docs/site/Decorators.md (hunk 3 above)
Comment (dhmlau) [reply docs/site/Decorators.md]: "it is recommended"?  
It's a bit strange to me for using "we" in the official documentation.

File: docs/site/Decorators.md (hunk 1 above)
Comment (jannyHou) [reply docs/site/Decorators.md]: Interesting...seems people have different opinions on the link's branch. I use "a specific version" before but Kevin corrected me to use master....personally I agree with "a specific version".

Appreciate more opinion from team @strongloop/lb-next-dev , for the code links in documentation, would you point them to `master` branch or a specific version?

Comment (dhmlau) [reply docs/site/Decorators.md]: It is often that the line numbers change over time, so I'd avoid pointing to the source code, no matter which branch it is pointing to.    My question to you, @jannyHou, would be.. what do we want to show to the users that we cannot show in a more static document, like API docs?

File: docs/site/Decorators.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (shimks) [reply docs/site/Decorators.md]: sure

File: docs/site/Decorators.md [hunk 15]
```diff
@@ -132,56 +132,183 @@ accordingly or do a composition of them:
...
+    description: 'a user',
+    required: true,
+    content: {
+      'application/json': {...schemaSpec},
+      'application/text': {...schemaSpec},
+    },
+  }
+  ```
+
+  When the decorated argument is a custom type or a model, we recommend you define
```
Comment (shimks) [inline docs/site/Decorators.md]: `If`, not `When`. I'd also elaborate why we 'recommend' users to decorate the model

File: docs/site/Decorators.md [hunk 16]
```diff
@@ -132,56 +132,183 @@ accordingly or do a composition of them:
...
+    @property()
+    lastname: string
+    @property()
+    address: Address
+  }
+  ```
+  *To learn more about decorating models and the corresponding OpenAPI schema, please check*
+  *[model decorators](#model-decorators).*
+
+  Then a typical usage of `@requestBody` is specifying a parameter's type as the model class,
```
Comment (shimks) [inline docs/site/Decorators.md]: Grammatically, this sentence is incomplete. I expect a sentence listing out the condition using 'if' before the sentence using 'then'

File: docs/site/Decorators.md [hunk 17]
```diff
@@ -132,56 +132,183 @@ accordingly or do a composition of them:
...
+  }
+  ```
+
+  For the simplest use case, you can leave the input of `@requestBody` empty, 
+  since we automatically detect the type of `user`, and generate the corresponding schema for it. 
+  The default content type is set to be `application/json`.
+
+  You can also customize the generated `requestBody` specification in 3 ways:
+
+  * add fields `description` and `required`
```
Comment (shimks) [inline docs/site/Decorators.md]: I'd like to see a more succinct organization of use cases of `requestBody`. Quoting from #1064 for inspiration:
> single content-type

    schema automatically generated with application/json
        with or without optional properties
    schema automatically generated with explicitly set content-type
        with or without optional properties
    schema set by user

File: docs/site/Decorators.md [hunk 18]
```diff
@@ -132,56 +132,183 @@ accordingly or do a composition of them:
...
       ) {}
     }
   ```
 
-  In the first example, we apply multiple decorators to a single declaration.
-  The order in which the decorators are called is important because a `@param`
-  decorator must be applied after an operation decorator.
-  To learn more about TypeScript decorator composition,
-  refer to [TypeScript Decorator Documentation](https://www.typescriptlang.org/docs/handbook/decorators.html)
+  You can find the specific usage in [Writing Controller methods](controllers.md#writing-controller-methods)
```
[the remark this answers is not in the document]
Comment (shimks) [reply docs/site/Decorators.md]: Also, the link is still dead

File: docs/site/Decorators.md [hunk 19]
```diff
@@ -132,56 +132,183 @@ accordingly or do a composition of them:
...
-  - in: one of the following values: `query`, `header`, `path`, `formData`, `body`
-  - type: one of the following values:  `string`, `number`, `boolean`, `integer`
-  - name: a `string`, name of the parameter
+  Syntax: see [API documentation](https://github.com/strongloop/loopback-next/blob/master/packages/openapi-v3/src/request-body-decorator.ts#L20-L79)
+
+  `@requestBody()` is applied to a controller method parameter to generate OpenAPI requestBody specification for it.
+  
+  *Only one parameter can be decorated by `@requestBody` per controller method.*
+
+  A typical OpenAPI requestBody spec contains properties `description`, `required`, and `content`:
```
Comment (shimks) [inline docs/site/Decorators.md]: Link here please https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#requestBodyObject

File: docs/site/Decorators.md [hunk 20]
```diff
@@ -132,56 +132,183 @@ accordingly or do a composition of them:
...
+    @put('/Users/{id}')
+    async replaceUser(
+      @param.path.string('id') id: string,
+      @requestBody() user: User
+    ) {}
+  }
+  ```
+
+  For the simplest use case, you can leave the input of `@requestBody` empty, 
+  since we automatically detect the type of `user`, and generate the corresponding schema for it.
```
Comment (shimks) [inline docs/site/Decorators.md]: no comma needed here

File: docs/site/Decorators.md [hunk 21]
```diff
@@ -132,56 +132,183 @@ accordingly or do a composition of them:
...
+  class UserController {
+    @put('/Users/{id}')
+    async replaceUser(
+      @param.path.string('id') id: string,
+      @requestBody() user: User
+    ) {}
+  }
+  ```
+
+  For the simplest use case, you can leave the input of `@requestBody` empty,
```
Comment (shimks) [inline docs/site/Decorators.md]: no comma between `'@requestBody' empty` and 'since we automatically'

File: docs/site/Decorators.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (jannyHou) [reply docs/site/Decorators.md]: @shimks I checked with Biniam, according to https://github.com/benbalter/jekyll-relative-links "controllers.md#writing-controller-methods" should be the correct format.
Do you have a valid example that I can look at?

File: docs/site/Decorators.md (hunk 17 above)
Comment (jannyHou) [reply docs/site/Decorators.md]: @shimks Sorry I don't quite agree with you here, the simplest example and 3 customized scenarios covered the following use cases:

```
schema automatically generated with application/json
    with or without optional properties
schema automatically generated with explicitly set content-type
    with or without optional properties
schema set by user
```

IMO there are three items can be customized:
- optional properties('description', 'required')
- content type
- schema spec

And that's how I classify the use cases.

File: docs/site/Decorators.md (hunk 15 above)
Comment (jannyHou) [reply docs/site/Decorators.md]: I think you didn't understand the context here. 
As I have mentioned:

```
If the decorated argument is a custom type or a model, we recommend you define 
  a model class with `@model` and `@property`

//  some code snippet

Then a typical usage of `@requestBody` is specifying a parameter's type as the model class,
  and applying the decorator on it:
```

The contents are consistent.

File: docs/site/Decorators.md [hunk 22]
```diff
@@ -132,56 +132,185 @@ accordingly or do a composition of them:
...
 
-  or
+  Writing the whole parameter specification is tedious, so we've created shortcuts to define
+  the params with the pattern `@param.${in}.${type}(${name})`:
+
+  - in: The parameter location. It can be one of the following values: `query`, `header`, `path`.
+  - type: A [common name of OpenAPI primitive data type](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#data-types). 
+  - name: Name of the parameter. It should be a `string`.
+
+  Take specifying a data from query as an example, a list of available shortcuts for can
```
Comment (shimks) [inline docs/site/Decorators.md]: I guess I'm confused as to what this sentence is trying to say. Did you possibly mean to say `Taking` instead of `Take`?

File: docs/site/Decorators.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (shimks) [reply docs/site/Decorators.md]: Oh, the PR might not have been rebased, that could be why the link is dead. I'd just advise to test out all links before pushing it upstream

File: docs/site/Decorators.md [hunk 23]
```diff
@@ -132,56 +132,185 @@ accordingly or do a composition of them:
...
+    @put('/Users/{id}')
+    async replaceUser(
+      @param.path.string('id') id: string,
+      @requestBody() user: User
+    ) {}
+  }
+  ```
+
+  For the simplest use case, you can leave the input of `@requestBody` empty
+  since we automatically detect the type of `user`, and generate the corresponding schema for it.
```
Comment (shimks) [inline docs/site/Decorators.md]: no comma

File: docs/site/Decorators.md (hunk 16 above)
Comment (shimks) [reply docs/site/Decorators.md]: I get what you're trying to say now with this section. Here's how I would word it based on the context you've given:
```
In order to use 'requestBody', the parameter type it's decorating needs to have its model decorated with '@model' and '@property'

CODE EXAMPLE

This allows type information of the model to be visible to the spec generator and '@requestBody' can finally be used on the parameter:
CODE EXAMPLE
```

File: docs/site/Decorators.md (hunk 1 above)
Comment (jannyHou) [reply docs/site/Decorators.md]: @dhmlau My link points to the tsdoc of source code :( I agree that this is the last choice...while could not think of a better idea as workaround

[the remark this answers is not in the document]
Comment (jannyHou) [reply docs/site/Decorators.md]: sorry replying late, there is a bug in tsdoc so I have to take this workaround, details see discussion:
https://github.com/strongloop/loopback-next/pull/1145#discussion_r175446623

[the remark this answers is not in the document]
Comment (jannyHou) [reply docs/site/Decorators.md]: @virkt25 BTW the fix for it is tracked in https://github.com/strongloop/loopback-next/issues/584#issuecomment-371544654. I added a comment there to remind myself update doc after it's fixed.

File: docs/site/Decorators.md [hunk 24]
```diff
@@ -141,58 +141,181 @@ For more usage, refer to
...
+    description: 'a user',
+    required: true,
+    content: {
+      'application/json': {...schemaSpec},
+      'application/text': {...schemaSpec},
+    },
+  }
+  ```
+
+  In order to use 'requestBody', the parameter type it's decorating needs to have its model decorated with '@model' and '@property'
```
Comment (b-admike) [inline docs/site/Decorators.md]: nitpick: suggest to use backtick ` for @model and @property instead of single quote.

File: docs/site/Decorators.md [hunk 25]
```diff
@@ -141,58 +141,181 @@ For more usage, refer to
...
+  ```
+  *To learn more about decorating models and the corresponding OpenAPI schema, please check
+  [model decorators](#model-decorators).*
+
+  This allows type information of the model to be visible to the spec generator so that `@requestBody` can be used on the parameter:
+
+  ```ts
+  import {User} from '../models/user'
+  import {put} from '@loopback/rest'
+  // in file controller.ts
```
Comment (b-admike) [inline docs/site/Decorators.md]: nitpick: suggest our convention of `/src/controllers/user.controller.ts` here

File: docs/site/Decorators.md (hunk 24 above)
Comment (jannyHou) [reply docs/site/Decorators.md]: good catch!

**Comment by shimks** [comment]
Is this the last of the task for #753? If it is, make sure to check it off in the milestone (#937)
